Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,68 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+        private readonly SemaphoreSlim actionsAvailable = new SemaphoreSlim(0);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!cancellation.IsCancellationRequested || !durableActions.IsEmpty || !transientActions.IsEmpty)
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
+                            actionsAvailable.Wait(cancellation.Token);
+                    }
+                    catch (OperationCanceledException)
+                    {
                     }
                     catch (Exception exception)
                     {
                         logger.ErrorException(string.Empty, exception);
                     }
-                }
             }) {Name = "EasyNetQ consumer dispatch thread", IsBackground = configuration.UseBackgroundThreads};
             thread.Start();
         }
 
         public void QueueAction(Action action)
+        {
+            QueueAction(action, false);
+        }
+
+        public void QueueAction(Action action, bool surviveDisconnect)
         {
             Preconditions.CheckNotNull(action, "action");
-            queue.Add(action);
+
+            if(cancellation.IsCancellationRequested)
+                throw new InvalidOperationException("Consumer dispatcher is stopping or already stopped");
+
+            if (surviveDisconnect)
+                durableActions.Enqueue(action);
+            else
+                transientActions.Enqueue(action);
+
+            actionsAvailable.Release();
```

Comment:
And yet I must admit that this option contains a serious flaw. A semaphore can increase its counter up to overflow, since `Release` happens every time a delegate is placed in a queue, and a `Wait` only if the both queues are empty. In addition, it will also give rise to some unnecessary work before a real lock, when the counter will be sequentially reset from the accumulated value to zero. I suggest using `ManualResetEventSlim` instead of `SemaphoreSlim`. It also has `Wait` method with cancellation token and more suited to our case.

Reply:
Agree. It was a quit buggy implementation, sorry :(

I have changed implementation again and hope it works fine now.